src/librustc_borrowck/borrowck/check_loans.rs

```diff
@@ -745,6 +745,26 @@ impl<'a, 'tcx> CheckLoanCtxt<'a, 'tcx> {
                                                      use_kind, lp_base);
             }
             LpExtend(ref lp_base, _, LpInterior(InteriorField(_))) => {
+                match lp_base.to_type().sty {
+                    ty::ty_struct(def_id, _) | ty::ty_enum(def_id, _) => {
+                        if ty::has_dtor(self.tcx(), def_id) {
+                            // In the case where the owner implements drop, then
+                            // the path must be initialized to prevent a case of
+                            // partial reinitialization
```

Comment:
How about this: Keep your code the way it is, but throw in a FIXME here and above `fn check_if_path_is_moved` noting that the body of `check_if_path_is_moved` is largely the same as the code you put into this function.  That way, if someone does this same thing again, maybe we will then be inspired to refactor the code over what error message to report.

(That, or you could add in such a refactoring yourself, as a followup commit post squash.)